source/tutorials/callpackage.md

```diff
@@ -0,0 +1,283 @@
+
+Things get more interesting when changing `default.nix` to make use of this new argument.
+Pass the parameter `audience` in the second argument to `callPackage`:
+
+```{code-block} diff
+:caption: default.nix
+ let
+   pkgs = import <nixpkgs> { };
+ in
+ {
+   hello = pkgs.callPackage ./hello.nix { };
++  hello-people = pkgs.callPackage ./hello.nix { audience = "people"; };
+ }
+```
+
+This attribute is passed on to the argument of the function defined in `hello.nix`:
+The same syntax can also be used to explicitly set the automatically discovered arguments, such as `writeShellScriptBin`, but that doesn't make sense here.
+
+Try it out:
+
+```shell-session
+$ nix-build -A hello-people
+$ ./result/bin/hello
+Hello, people!
+```
+
+This pattern is used widely in Nixpkgs:
+For example, functions which represent Go programs often have a parameter `buildGoModule`.
+It is common to find expressions like `callPackage ./go-program.nix { buildGoModule = buildGo116Module; }` to enforce a certain Go compiler version.
+Nixpkgs is therefore not simply a huge library of pre-configured packages, but a collection of functions – package *recipes* – for customising packages and even entire ecosystems (for example "All Python packages using my custom interpreter") on the fly without duplicating code.
+
+# Overrides
+
+`callPackage` adds more convenience by adding an attribute to the derivation it returns: the `override` function.
+This means that, as a consequence of parameterising builds with `callPackage`, you can also change the value of these parameters _after the fact_ using the returned derivation's `override` function.
+
+Add a third attribute `hello-folks` to `default.nix` and set it to `hello.override` called with a new value for `audience`:
+
+```{code-block} diff
+:caption: default.nix
+ let
+   pkgs = import <nixpkgs> { };
+ in
+-{
++rec {
+   hello = pkgs.callPackage ./hello.nix { };
+   hello-people = pkgs.callPackage ./hello.nix { audience = "people"; };
++  hello-folks = hello.override { audience = "folks"; };
+ }
+```
+
+:::{note}
+The resulting attribute set is now recursive (by the keyword `rec`).
+That is, attribute values can refer to names from within the same attribute.
+:::
+
+`override` passes `audience` to the original function in `hello.nix` - it *overrides* whatever arguments have been passed in the original `callPackage` that produced the derivation `hello`.
+All the other parameters will remain the same.
+This is especially useful and can be often found on packages that provide many options to customise a package.
+
+Building `hello-folks` attribute and running the resulting executable will again produce a new version of the script:
+
+```shell-session
+$ nix-build -A hello-folks
+$ ./result/bin/hello
+Hello, folks!
+```
+
+A real-world example is the [`neovim`](https://search.nixos.org/packages?show=neovim) package recipe, which has has overrideable arguments such as `extraLuaPackages`, `extraPythonPackages`, or `withRuby`.
```

Comment:
Important to point out is that these parameters are only discoverable by looking at the source file, which people can get to by clicking the "Source" link on search.nixos.org